File: 1-D Dynamic Programming/790. Domino and Tromino Tiling/solution.py
class Solution:
    def numTilings(self, n: int) -> int:
        MOD = 1000000007
        # dp[i]: # of ways to tile 2 x i board
        dp = [0] * (n+1)
        
        # base case
        dp[0] = 1
        dp[1] = 1
        if n == 1: return 1
        dp[2] = 2
        if n == 2: return 2

        # dp[i] = dp[i-1] + dp[i-2] + 2 * (dp[i-3] + dp[i-4] + ... dp[0])
        presumDp = dp[1]
        for i in range(3, n+1):
            dp[i] = dp[i-1] + dp[i-2] + 2*presumDp
            dp[i] %= MOD
            presumDp += dp[i-2]
        return dp[n]

File: 1-D Dynamic Programming/790. Domino and Tromino Tiling/solution.cpp
// https://www.youtube.com/watch?v=S-fUTfqrdq8&ab_channel=HuaHua
using LL = long long;

class Solution {
public:
    int mod = 1e9 + 7;
    int numTilings(int n) {
        vector<vector<LL>> dp(n+1, vector<LL>(2, 0));
        dp[0][0] = dp[1][0] = 1;
        for (int i=2; i<= n; i++) {
            dp[i][0] = (dp[i-1][0] + dp[i-2][0] + 2 * dp[i-1][1]) % mod;
            dp[i][1] = (dp[i-2][0] + dp[i-1][1]) % mod;
        }
        return dp[n][0];
    }
};
